libs/domain/picking/offline/data-plugin.ts

```diff
@@ -0,0 +1,107 @@
+import { OauthService } from '@spryker-oryx/auth';
+import {
+  AppEnvironment,
+  ExecPlugin,
+  InjectionPlugin,
+} from '@spryker-oryx/core';
+import { Injector } from '@spryker-oryx/di';
+import { DexieIndexedDbService } from '@spryker-oryx/indexed-db';
+import {
+  combineLatest,
+  map,
+  Observable,
+  of,
+  Subscription,
+  switchMap,
+  withLatestFrom,
+} from 'rxjs';
+import { PickingListEntity, PickingProductEntity } from './entities';
+import { PickingListOnlineAdapter } from './services';
+
+export class OfflineDataPlugin extends ExecPlugin {
+  protected subscription?: Subscription;
+
+  constructor() {
+    super((app) => {
+      const injector = app!.requirePlugin(InjectionPlugin).getInjector();
+
+      const env = injector.inject(AppEnvironment, {} as AppEnvironment);
+
+      const authService = injector.inject(OauthService);
+
+      this.subscription = authService
+        .isAuthenticated()
+        .pipe(
+          switchMap((authenticated) => {
+            if (authenticated) {
+              return this.clearDb(injector).pipe(
+                switchMap(() => this.populateDb(injector))
+              );
+            }
+            return of(undefined);
+          })
+        )
+        .subscribe();
+    });
+  }
+
+  destroy(): void {
+    this.subscription?.unsubscribe();
+  }
+
+  protected clearDb(injector: Injector): Observable<void> {
+    const dexieIdbService = injector.inject(DexieIndexedDbService);
+
+    return combineLatest([
+      dexieIdbService.getStore(PickingListEntity),
+      dexieIdbService.getStore(PickingProductEntity),
+    ]).pipe(
+      switchMap(async ([pickingStore, productStore]) => {
+        await pickingStore.clear();
+        await productStore.clear();
+        return undefined;
+      })
+    );
+  }
+
+  protected populateDb(injector: Injector): Observable<void> {
+    const onlineAdapter = injector.inject(PickingListOnlineAdapter);
+    const dexieIdbService = injector.inject(DexieIndexedDbService);
+
+    return onlineAdapter.get({}).pipe(
+      map((pl) => {
+        const productIds = new Set<string>();
+
+        return {
+          pickingLists: pl,
+          products: pl
+            .map((pickingList) => pickingList.items.map((item) => item.product))
+            .flat()
+            .filter((product) => {
```

Comment:
```
.filter((product, i, arr) => {
  return arr.findIndex(p => p.id === product.id) === i;
}),
```

Reply:
This doesn't work. And it's missing `productIds.add(product.id)`.

Reply:
Why? This is doing the same thing, filter duplications, but in one string, without additional Set's and etc. `productIds` is not needed anymore

Reply:
This is going to be a much slower performance wise as you are doing loop in a loop to find duplicates while Set is O(1) which is much faster.
So I would suggest to revert to a previous implementation assuming that we are downloading the whole set of picking lists in one request.

Reply:
So, do you think that just reading data already stored in memory is less performance then writing new data to the memory during updating the Set and searching in it? I think the difference is not significant.

Reply:
Absolutely, you will only write once, but you will read every time the same array for every item.
The difference is literally exponential.